ProcessExtensions/ProcessExtensions.cs

```diff
@@ -1,233 +1,98 @@
-                    current += arrayElementSize;
-
-                    if (si.State == WTS_CONNECTSTATE_CLASS.WTSActive)
+                    if (sessionInfo[i].State == WTS_CONNECTSTATE_CLASS.WTSActive)
                     {
-                        activeSessionId = si.SessionID;
+                        activeSessionId = sessionInfo[i].SessionId;
+                        break;
                     }
                 }
 
                 WTSFreeMemory(pSessionInfo);
             }
 
-            // If enumerating did not work, fall back to the old method
             if (activeSessionId == INVALID_SESSION_ID)
             {
                 activeSessionId = WTSGetActiveConsoleSessionId();
+
+                if (activeSessionId == INVALID_SESSION_ID)
+                {
+                    return false;
+                }
             }
 
             if (WTSQueryUserToken(activeSessionId, ref hImpersonationToken) != 0)
             {
-                // Convert the impersonation token to a primary token
-                bResult = DuplicateTokenEx(hImpersonationToken, 0, IntPtr.Zero,
-                    (int)SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, (int)TOKEN_TYPE.TokenPrimary,
-                    ref phUserToken);
+                var hImpersonationTokenSafe = new SafeFileHandle(hImpersonationToken, true);
+
+                bResult = DuplicateTokenEx(hImpersonationTokenSafe, 0, null, SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, TOKEN_TYPE.TokenPrimary, out phUserToken);
 
                 CloseHandle(hImpersonationToken);
             }
 
             return bResult;
         }
 
-        public static bool StartProcessAsCurrentUser(string appPath, string cmdLine = null, string workDir = null, bool visible = true)
+        /// <summary>
+        /// Starts a process as the current active user.
+        /// </summary>
+        /// <param name="appPath">The path to the application to start.</param>
+        /// <param name="cmdLine">The command line arguments for the process.</param>
+        /// <param name="workDir">The working directory for the process.</param>
+        /// <param name="visible">Whether the process window should be visible.</param>
+        /// <returns>True if the process starts successfully; otherwise, false.</returns>
+        /// <exception cref="ProcessCreationException">Thrown when the process could not be started.</exception>
+        public unsafe static bool StartProcessAsCurrentUser(string appPath, string cmdLine = null, string workDir = null, bool visible = true)
         {
-            var hUserToken = IntPtr.Zero;
-            var startInfo = new STARTUPINFO();
+            SafeFileHandle hUserToken = null;
+            var startInfo = new STARTUPINFOW();
             var procInfo = new PROCESS_INFORMATION();
-            var pEnv = IntPtr.Zero;
+            void* pEnv = null;
```

Comment:
Curious why you made this change (and the others like it)? My understanding (from years ago) was that `IntPtr.Zero` is the recommended way to initialize pointers to `0`, and using `null` feels a bit odd to me. Have the conventions shifted towards this?

Reply:
In C#, within an unsafe code block, it's common to use null for uninitialized void* pointers. This aligns with the syntax for pointers in C# when unsafe context is enabled. Here's a brief comparison to illustrate:

Using IntPtr requires additional steps:

```csharp
unsafe
{
    IntPtr ptr = IntPtr.Zero; // Safe handle initialization
    void* voidPointer = ptr.ToPointer(); // Conversion to void*
    SomeMethodExpectingVoidPointer(voidPointer);
}
```

Directly using null is more straightforward:

```csharp
unsafe
{
    void* voidPointer = null; // Directly initialize with null
    SomeMethodExpectingVoidPointer(voidPointer);
}
```

The method in question expects a void* type. Since IntPtr is a safe handle and the method expects an actual pointer, using null for void* is simpler and avoids the extra step of conversion. This is standard in unsafe C# contexts and makes the intent clear that we are dealing with a pointer that is not yet pointing to any memory location.